File: web/di_core/LakeHouse/Service/LakeJobService.ts
import { Inject } from 'typescript-ioc';
import { LakeJob } from '@core/LakeHouse/Domain/LakeJob/LakeJob';
import { JobId } from '@core/domain';
import { ListingResponse } from '@core/DataIngestion';
import { LakeJobRepository } from '@core/LakeHouse/Repository/LakeJobRepository';
import { LakeJobResponse } from '@core/LakeHouse/Domain/LakeJob/LakeJobResponse';
import { LakeJobHistory } from '@core/LakeHouse/Domain/LakeJob/LakeJobHistory';
import { ListingRequest } from '@core/LakeHouse/Domain/Request/ListingRequest/ListingRequest';
import { ForceMode } from '@core/LakeHouse/Domain/LakeJob/ForceMode';

export abstract class LakeJobService {
  abstract create(job: LakeJob): Promise<LakeJob>;
  abstract get(jobId: JobId): Promise<LakeJobResponse>;
  abstract list(request: ListingRequest): Promise<ListingResponse<LakeJobResponse>>;
  abstract listHistory(request: ListingRequest): Promise<ListingResponse<LakeJobHistory>>;
  abstract delete(jobId: JobId): Promise<boolean>;
  abstract update(job: LakeJob): Promise<boolean>;
  abstract forceRun(jobId: JobId, date: number, mode: ForceMode): Promise<boolean>;
  abstract cancel(jobId: JobId): Promise<boolean>;
}
export class LakeJobServiceImpl extends LakeJobService {
  constructor(@Inject private repository: LakeJobRepository) {
    super();
  }

  create(job: LakeJob): Promise<LakeJob> {
    return this.repository.create(job);
  }

  delete(jobId: JobId): Promise<boolean> {
    return this.repository.delete(jobId);
  }

  get(jobId: JobId): Promise<LakeJobResponse> {
    return this.repository.get(jobId);
  }

  list(request: ListingRequest): Promise<ListingResponse<LakeJobResponse>> {
    return this.repository.list(request);
  }
  listHistory(request: ListingRequest): Promise<ListingResponse<LakeJobHistory>> {
    return this.repository.listHistory(request);
  }

  update(job: LakeJob): Promise<boolean> {
    return this.repository.update(job);
  }

  forceRun(jobId: JobId, date: number, mode: ForceMode): Promise<boolean> {
    return this.repository.forceRun(jobId, date, mode);
  }

  cancel(jobId: JobId): Promise<boolean> {
    return this.repository.cancel(jobId);
  }
}

File: job-scheduler/src/main/scala/datainsider/lakescheduler/service/LakeJobService.scala
package datainsider.lakescheduler.service

import com.twitter.util.Future
import datainsider.client.domain.ThriftImplicit._
import datainsider.client.exception.BadRequestError
import datainsider.client.service.ProfileClientService
import datainsider.client.util.TimeUtils
import datainsider.jobscheduler.domain.Ids.JobId
import datainsider.jobscheduler.domain.job.JobStatus.JobStatus
import datainsider.jobscheduler.domain.request.PaginationResponse
import datainsider.lakescheduler.domain.job.LakeJob
import datainsider.lakescheduler.domain.request.{ListLakeJobRequest, UpdateLakeJobRequest}
import datainsider.lakescheduler.domain.response.LakeJobResponse
import datainsider.lakescheduler.repository.LakeJobRepository

import javax.inject.Inject

trait LakeJobService {
  def create(orgId: Long, job: LakeJob): Future[LakeJobResponse]

  def delete(orgId: Long, id: JobId): Future[Boolean]

  def update(orgId: Long, request: UpdateLakeJobRequest): Future[Boolean]

  def get(orgId: Long, jobId: JobId): Future[LakeJobResponse]

  def list(orgId: Long, request: ListLakeJobRequest): Future[PaginationResponse[LakeJobResponse]]

  def getWith(orgId: Long, jobStatus: List[JobStatus], from: Int, size: Int): Future[Seq[LakeJob]]
}

case class LakeJobServiceImpl @Inject() (
    lakeJobRepository: LakeJobRepository,
    profileClientService: ProfileClientService
) extends LakeJobService {

  override def create(orgId: Long, job: LakeJob): Future[LakeJobResponse] = {
    for {
      jobId <- lakeJobRepository.insert(orgId, job.customCopy(nextRunTime = TimeUtils.calculateNextRunTime(job.scheduleTime, None)))
      job <- lakeJobRepository.get(orgId, jobId)
    } yield job match {
      case None        => throw new InternalError("Fail to get job")
      case Some(value) => toLakeJobResponse(value)
    }
  }.flatten

  override def delete(orgId: Long, id: JobId): Future[Boolean] = {
    lakeJobRepository.delete(orgId, id)
  }

  override def update(orgId: Long, request: UpdateLakeJobRequest): Future[Boolean] = {
    lakeJobRepository.update(orgId, request.job.customCopy(nextRunTime = TimeUtils.calculateNextRunTime(request.job.scheduleTime, None)))
  }

  override def get(orgId: Long, jobId: JobId): Future[LakeJobResponse] = {
    lakeJobRepository.get(orgId, jobId).map {
      case None      => throw BadRequestError(s"No job was found for id = $jobId.")
      case Some(job) => toLakeJobResponse(job)
    }
  }.flatten

  override def list(orgId: Long, request: ListLakeJobRequest): Future[PaginationResponse[LakeJobResponse]] = {
    for {
      jobs <- lakeJobRepository.list(orgId, request.keyword, request.from, request.size, request.sorts)
      responses <- toLakeJobResponses(orgId, jobs)
      total <- lakeJobRepository.count(orgId, request.keyword)
    } yield PaginationResponse(data = responses, total = total)
  }

  override def getWith(orgId: Long, jobStatus: List[JobStatus], from: Int, size: Int): Future[Seq[LakeJob]] = {
    lakeJobRepository.getWith(orgId, jobStatus, from, size)
  }

  private def toLakeJobResponses(orgId: Long, jobs: Seq[LakeJob]): Future[Seq[LakeJobResponse]] = {
    val creatorIds: Seq[String] = jobs.map(_.creatorId).distinct
    for {
      creatorProfiles <- profileClientService.getUserProfiles(orgId, creatorIds)
      responses = jobs.map(job => LakeJobResponse(job, creatorProfiles.get(job.creatorId).map(_.toShortUserProfile)))
    } yield responses
  }

  private def toLakeJobResponse(job: LakeJob): Future[LakeJobResponse] = {
    profileClientService.getUserProfile(job.orgId, job.creatorId).map(userProfile => LakeJobResponse(job, userProfile.map(_.toShortUserProfile)))
  }
}
